spacy/language.py

```diff
@@ -999,6 +997,52 @@ def from_bytes(self, bytes_data, exclude=tuple(), disable=None, **kwargs):
         return self
 
 
+class component(object):
+    """Decorator for pipeline components. Can decorate both function components
+    and class components and will automatically register components in the
+    Language.factories. If the component is a class and needs access to the
+    nlp object or config parameters, it can expose a from_nlp classmethod
+    that takes the nlp object and **cfg arguments and returns the initialized
+    component.
+    """
+
+    # NB: This decorator needs to live here, because it needs to write to
+    # Language.factories. All other solutions would cause circular import.
+
+    def __init__(self, name=None, assigns=tuple(), requires=tuple(), retokenizes=False):
+        """Decorate a pipeline component.
+
+        name (unicode): Default component and factory name.
+        assigns (list): Attributes assigned by component, e.g. `["token.pos"]`.
+        requires (list): Attributes required by component, e.g. `["token.dep"]`.
+        retokenizes (bool): Whether the component changes the tokenization.
+        """
+        self.name = name
+        self.assigns = validate_attrs(assigns)
+        self.requires = validate_attrs(requires)
+        self.retokenizes = retokenizes
+
+    def __call__(self, *args, **kwargs):
+        obj = args[0]
+        args = args[1:]
+        factory_name = self.name or util.get_component_name(obj)
+        obj.name = factory_name
+        obj.factory = factory_name
+        obj.assigns = self.assigns
+        obj.requires = self.requires
+        obj.retokenizes = self.retokenizes
```

Comment:
It would be nice to have a design that didn't even modify the object. There could be unexpected results if people have an attribute in their class that conflicts. `name` seems especially likely to cause problems.

Perhaps if we only assigned one attribute, `component_info`, that might be better?

Alternatively we could store the information separately in a global variable, and have helper methods like `get_factory_name(component)` which would look it up. We're already using `getattr` to access the attribute anyway, so it doesn't really seem worse to lookup the value with a function.

Reply:
> `name` seems especially likely to cause problems.

This has already been the default since we've introduced the new style of components, though. We're not writing to it, but we're expecting class components to have an attribute `name`.

> There could be unexpected results if people have an attribute in their class that conflicts.

I get what you mean, but the idea here is that these are some of the few hard-coded attributes we expect. We also need class components to be able to provide this data independently from the decorator – see the `cdef` classes like `EntityRecognizer` for instance, which can't be decorated and defines these attributes and registers itself in the factories instead.

> We're already using getattr to access the attribute anyway, so it doesn't really seem worse to lookup the value with a function.

I guess we could have a function `register_component` that takes the factory and the other configuration (`requires`, `assigns` etc.). This can be called from within the decorator, but also easily outside of it (e.g. to add the Cython classes).